### ISSUE TITLE
Can we use backward within the forward function of a criterion?

### ISSUE THREAD

**Issue by ashwath98** [facebookresearch/fairseq#4618]
#### What is your question?
Can we use a loss.backward() call within a criterion in fairseq? I am trying to write an adversarial version of the cross entropy loss. 
#### Code (Forward Function)
       `sample2=copy.deepcopy(sample) 
        sample2['net_input']['source']['video'].requires_grad=True
        net_output = model(**sample2["net_input"])
        loss_dummy, _ = self.compute_loss(model, net_output, sample2, reduce=reduce)
        sample2['net_input']['source']['video'].retain_grad()
        model.zero_grad()
        loss_dummy.backward()
        grad=sample2['net_input']['source']['video'].grad
        sign_data_grad = grad.sign()
        sample2['net_input']['source']['video']=sample2['net_input']['source']['video']+self.epsilon*sign_data_grad
        model.zero_grad()
        net_output_adv = model(finetune=False,**sample2["net_input"])
        loss_adv, nll_loss_adv = self.compute_loss(model, net_output_adv, sample2, reduce=reduce)
        net_output = model(finetune=False,**sample["net_input"])
        loss_clean, nll_loss = self.compute_loss(model, net_output, sample, reduce=reduce)
        sample_size = (
            sample["target"].size(0) if self.sentence_avg else sample["ntokens"]
        )
        loss=self.alpha*loss_clean+(1-self.alpha)*loss_adv`

#### What have you tried?
The line loss_dummy.backward() throws this error : Runtime error : **Element 0 of tensor does not require grad and does not have a grad function. (During a fair-seq hydra script)**

I tried  using the criterion in a python script, and it works perfectly fine, which made me assume it could be something to do with fairseq under the hood.

#### What's your environment?

 - fairseq Version (e.g., 1.0 or main): commit from https://github.com/facebookresearch/av_hubert
 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): Linux
 - How you installed fairseq (`pip`, source) : source 
 - Build command you used (if compiling from source):(Instructions from https://github.com/facebookresearch/av_hubert)
 - Python version: 3.8
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

**Comment by ashwath98** [comment]
resolved
